adam/visualization/positioning.py

```diff
@@ -602,5 +710,168 @@ def overlap_penalty(min_max_overlaps: torch.Tensor) -> torch.Tensor:
         return overlap_distance.max() * -1 * COLLISION_PENALTY
 
 
+class RelationalPenalty(nn.Module):  # type: ignore
+    """ Model that penalizes boxes for not adhering to relational (distance and direction)
+        constraints with other boxes
+    """
+
+    def __init__(self):  # pylint: disable=useless-super-delegation
+        super().__init__()
+
+    def forward(self, *inputs):  # pylint: disable=arguments-differ
+        bounding_box_1: AxisAlignedBoundingBox = inputs[0]
+        bounding_box_2: AxisAlignedBoundingBox = inputs[1]
+        # adam objects store the constraints
+        adam_object_1: AdamObject = inputs[2]
+        adam_object_2: AdamObject = inputs[3]
+
+        print(f"{adam_object_1.name} positioned w/r/t {adam_object_2.name}")
+
+        # return 0 if object has no relative positions to apply
+        if not adam_object_1.relative_positions:
+            print(f"{adam_object_1.name} has no relative positioning constraints")
+            return torch.zeros(1)
+        if adam_object_2 not in adam_object_1.relative_positions:
+            print(f"{adam_object_1.name} not positioned relative to {adam_object_2.name}")
+            return torch.zeros(1)
+
+        if adam_object_2 in adam_object_1.relative_positions:
+            return RelationalPenalty.penalty(
+                bounding_box_1,
+                bounding_box_2,
+                adam_object_1.relative_positions[adam_object_2],
+            )
+
+    @staticmethod
+    def penalty(
+        box1: AxisAlignedBoundingBox,
+        box2: AxisAlignedBoundingBox,
+        box1_2_relation: InRegionDirectives,
+    ):
+        """
+        Assign a penalty for box1 if it does not comply with its relation to box2
+        Args:
+            box1:
+            box2:
+            box1_2_relation: relation of box1 to box2
+
+        Returns: Tensor(1,) with a penalty
+
+        """
+        # get direction that box 1 should be in w/r/t box 2
+        # TODO: allow for addressee directions
+        direction_vector = RelationalPenalty.direction_to_world(box1_2_relation.direction)
+
+        current_vector = box1.center - box2.center
```

Comment:
`current_vector` raises the question "which vector?" for the reader.  Would `current_direction_from_reference_to_object` be better?